File: Queue/Minimum K Consecutive Bit Flips/optimal_app.java
import java.util.*

//Time complexity :- O(n)
//Space complexity :- O(k)

// Approach :- 
// instead of storing all the n index information 
// we can store only k bit info as that is only what we need 
// we'll use deque to store that info 
// we'll add 1 if we flip the bit otherwise 0 always 


//Link :- https://www.geeksforgeeks.org/problems/minimum-number-of-k-consecutive-bit-flips--171650/1

class Solution {
    public int kBitFlips(int[] arr, int k) {
        int n = arr.length;
        int flipped = 0, res = 0;
        Deque<Integer> dq = new LinkedList<>();
        
        for(int i=0; i<n; ++i){
            // validity of previous flip is expired 
            if(i >= k){
                flipped -= dq.peekFirst();
                dq.pollFirst();
            }
            // flip is needed 
            if(arr[i] == (flipped % 2)){
                // if index does not have sufficient next bit then it's not possible 
                if(i+k > n){
                    return -1;
                }
                // add the condition to flips 
                res += 1;
                flipped++;
                dq.offerLast(1);
            }else{
                dq.offerLast(0);
            }
        }
        return res;
    }
}

File: Array/Basics/Express as Consecutive Number Sum/optimal_app.cpp
#include <bits/stdc++.h>
using namespace std;

//Time complextiy :- O(1)
// Space complexity :- O(1)

// Approach :- 
// if the number is power of 2 then it cannot be expressed as power of 2 


// Link :-  https://www.geeksforgeeks.org/problems/consecutive-numbers-for-sum3132/1

class Solution {
    public boolean isSumOfConsecutive(int n) {
        // base case 
        if(n == 1)
            return false;
        
        // is power of 2 
        if((n & (n-1)) == 0)
            return false;
        
        return true;
    }
}